### ISSUE TITLE
Mapping to `Uint8Array`

### ISSUE THREAD

**Issue by Liamolucko** [WebAssembly/component-model#66]
It seems like the component model's recommended way of representing bytes is `list<u8>`, which is reasonable. However, in the JavaScript binding, lists always map to JS arrays, which seems pretty inefficient for passing bytes around. Is there any plan for defining APIs which take/return `Uint8Array`s in the JS binding, whether that be with `list<u8>` or some other type?

(Side-note - I was originally going to ask about representing TypedArrays in general, but it doesn't really seem doable for anything other than `Uint8Array` and `Int8Array`, since WebAssembly is little endian and TypedArrays are native endian.)

**Comment by lukewagner** [comment]
Ah yes, great point!  I had been assuming this but forgot to write it down.  I think it even makes sense to pass `list<u16>` and `list<u32>` as typed arrays: because of the copy semantics of `list` (such that typed arrays don't retain a live view on a component's linear memory), there's a copying happening anyways and so, if an endianness transform was required, it could easily happen at the copy.

**Comment by Liamolucko** [comment]
> I think it even makes sense to pass `list<u16>` and `list<u32>` as typed arrays: because of the copy semantics of `list` (such that typed arrays don't retain a live view on a component's linear memory), there's a copying happening anyways and so, if an endianness transform was required, it could easily happen at the copy.

👍 Good point, I didn't think of that.

Would that also include `list<f32>` and `list<f64>`? On the one hand, that seems like something that's more likely to be consumed at a high level where you don't want typed arrays involved; on the other hand, typed arrays aren't that different to use than regular arrays, and it's pretty easy to convert between them. So, it might be worth it for the efficiency + binding directly to the couple of Web APIs that take `Float32Array`.

**Comment by lukewagner** [comment]
Yep, I think `list<float32>` and `list<float64>` make sense as typed arrays too.  (On the incoming direction, the `Float32Array`/`Float64Array` floats will need to be canonicalized.)
